### PR TITLE
feat: use custom handlers to pay for gas with erc-20

### PR DESCRIPTION
**Pull request by royvardhan**
Closes: #1782

### CODE REVIEW

File: examples/erc20_gas/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    let mut evm = MainEvm::new(
+        Context::builder()
+            .with_db(cache_db)
+            .modify_tx_chained(|tx| {
+                tx.caller = from;
+                tx.transact_to = TxKind::Call(token);
+                tx.data = encoded.into();
+                tx.value = U256::from(0);
+            }),
+        EthHandler::default(),
```
Comment (royvardhan) [inline examples/erc20_gas/src/main.rs:181]: @rakita So when I try to plug the custom handlers, it does not let me. As it expects the original handlers here with all the trait bounds. But the ones I have implemented also satisfies the trait bounds.

Comment (rakita) [reply examples/erc20_gas/src/main.rs:181]: Where is the implementation of your Handler?

Comment (royvardhan) [reply examples/erc20_gas/src/main.rs:181]: Here: https://github.com/royvardhan/revm/tree/1782-1/examples/erc20_gas/src/handlers

Comment (rakita) [reply examples/erc20_gas/src/main.rs:181]: Will need to check errors

Comment (royvardhan) [reply examples/erc20_gas/src/main.rs:181]: I am trying to do this: 
image

Comment (royvardhan) [reply examples/erc20_gas/src/main.rs:181]: One of the errors: ```
arguments to this function are incorrect
expected struct `EthValidation<Context<BlockEnv, TxEnv, CfgEnv, &mut CacheDB<WrapDatabaseAsync<AlloyDB<..., ..., ...>>>>, ...>`
   found struct `Erc20Validation`
```

Comment (rakita) [reply examples/erc20_gas/src/main.rs:181]: Hm it seems that EthHandler::new needs to be changed slightly.

Comment (royvardhan) [reply examples/erc20_gas/src/main.rs:181]: There is no `new` method in `[Handler](https://github.com/bluealloy/revm/blob/45581c451a440776fd59576d7b27c366b1528724/crates/handler/interface/src/handler.rs#L3)`. So I did this, still not supporting: 
image

Comment (royvardhan) [reply examples/erc20_gas/src/main.rs:181]: Due to this [MainEvm](https://github.com/bluealloy/revm/blob/45581c451a440776fd59576d7b27c366b1528724/crates/revm/src/evm.rs#L150) it is taking the default handler as `EthHandler`

Comment (FredCoen) [reply examples/erc20_gas/src/main.rs:181]: did you try creating a type alias for your handler like its done for op here https://github.com/bluealloy/revm/blob/45581c451a440776fd59576d7b27c366b1528724/crates/optimism/src/handler.rs#L40

and then do a CustomEvm type alias similar to MainEvm but replaces the default Ethhandlert.
and then go CustomEvm::new(..

Comment (royvardhan) [reply examples/erc20_gas/src/main.rs]: @FredCoen thanks man, fixed!

**Comment by royvardhan** [comment]
@rakita I think we are good

File: examples/erc20_gas/src/main.rs (hunk 1 above)
Comment (FredCoen) [reply examples/erc20_gas/src/main.rs]: awesome 💪💪

Comment (rakita) [reply examples/erc20_gas/src/main.rs]: Awesome. I will check it tomorrow

**Review by rakita — APPROVED** [review]
lgtm

**Comment by rakita** [comment]
@royvardhan can you merge with newest main, just so that the CI passes?

**Comment by royvardhan** [comment]
> @royvardhan can you merge with newest main, just so that the CI passes?

Yes ser. Will do

**Pull request merged**